pyteal/ast/subroutine.py

```diff
@@ -40,73 +49,149 @@ def __init__(
             )
 
-        for name, param in sig.parameters.items():
-            if param.kind not in (
-                Parameter.POSITIONAL_ONLY,
-                Parameter.POSITIONAL_OR_KEYWORD,
-            ):
-                raise TealInputError(
-                    "Function has a parameter type that is not allowed in a subroutine: parameter {} with type {}".format(
-                        name, param.kind
-                    )
-                )
-
-            if param.default != Parameter.empty:
-                raise TealInputError(
-                    "Function has a parameter with a default value, which is not allowed in a subroutine: {}".format(
-                        name
-                    )
-                )
-
-            expected_arg_type = self._validate_parameter_type(annotations, name)
-
-            self.expected_arg_types.append(expected_arg_type)
-            if expected_arg_type is ScratchVar:
-                self.by_ref_args.add(name)
+        # validate full signature takes following two arguments:
+        # - `signature`, which contains the signature of the python function.
+        #    NOTE: it contains all the arguments, we get type annotations from `annotations`.
+        # - `annotations`, which contains all available argument type annotations and return type annotation.
+        #    NOTE: `annotations` does not contain all the arguments,
+        #          an argument is not included in `annotations` if its type annotation is not available.
+        (
+            expected_arg_types,
+            by_ref_args,
+            abi_args,
+        ) = SubroutineDefinition._arg_types_and_by_refs(sig, annotations)
+        self.expected_arg_types: List[
+            Union[Type[Expr], Type[ScratchVar], abi.TypeSpec]
+        ] = expected_arg_types
+        self.by_ref_args: Set[str] = by_ref_args
+        self.abi_args: Dict[str, abi.TypeSpec] = abi_args
 
         self.implementation = implementation
-        self.implementationParams = sig.parameters
-        self.returnType = returnType
+        self.implementation_params = sig.parameters
+        self.return_type = return_type
 
         self.declaration: Optional["SubroutineDeclaration"] = None
         self.__name = self.implementation.__name__ if nameStr is None else nameStr
 
+    @staticmethod
+    def is_abi_class(arg_type: Type) -> bool:
+        return (
+            isclass(arg_type)
+            and arg_type is not abi.BaseType
+            and issubclass(arg_type, abi.BaseType)
+        )
+
+    @staticmethod
+    def is_abi_annotation(obj: Any) -> bool:
+        try:
+            abi.type_spec_from_annotation(obj)
+            return True
+        except TypeError:
+            return False
+
     @staticmethod
     def _validate_parameter_type(
         user_defined_annotations: dict, parameter_name: str
-    ) -> Type[Union[Expr, ScratchVar]]:
+    ) -> Union[Type[Union[Expr, ScratchVar]], abi.TypeSpec]:
```

Comment:
Is this equivalent to `Union[Type[Expr], Type[ScratchVar], abi.TypeSpec]`? If so I think the latter approach is more clear.

Reply:
yes, I wrote this a while ago, and I think in this implementation I did use the annotation you suggested here.